tests/test_ticks.c

```diff
@@ -0,0 +1,149 @@
+// Writes the 32bit value to COP0 register 9 (count)
+#define WRITE_TICKS(val) do { asm volatile("mtc0 %0,$9"::"r"(val)); } while(0)
+
+// Writes the 32bit value to COP0 register 11 (compare)
+#define WRITE_COMPARE(val) do { asm volatile("mtc0 %0,$11" ::"r"(val)); } while(0)
+
+#define ARM_INTERRUPT(compare) do { \
+	asm volatile(".align 4"); \
+	WRITE_TICKS(0x0); \
+	WRITE_COMPARE(compare); \
+} while(0)
+
+#define TEST(name, tests) do { \
+	const int NUM_CASES = sizeof(tests) / sizeof(tests[0]); \
+	for (int i=0; i < NUM_CASES; i++) { \
+		mock_tick = tests[i][1]; \
+		WRITE_TICKS(tests[i][0]); \
+		/* act */ \
+		name(tests[i][2]); \
+		READ_TICKS(ticks_0); \
+		ASSERT(state != Timeout, "Case: %u Test timed out. " #name " didn't finish on time (Ticks: %#10lX)", state, ticks_0); \
+		ASSERT(state == Test, "Case: %u " #name " finished too early (Ticks: %#10lX)", state, ticks_0); \
+		ASSERT(state <= Timeout, "Case: %u Unexpected state for " #name " (Ticks: %#10lX)", state, ticks_0); \
+		/* re-sync to frame */ \
+		while (state < Timeout); \
+		/* reset state */ \
+		state = Start; \
+	} \
+} while(0)
+
+static volatile uint32_t mock_tick = 0;
+static volatile enum state_t{ Start = 0, Test = 1, Timeout = 2 } state = -1;
+
+static void frame_callback() {
+	state++;
+
+	switch(state) {
+		case Test:
+			WRITE_TICKS(mock_tick);
+		break;
+		default:
+		break;
+	}
+}
+
+// Array of test cases in the form {initial, mock, wait}
+// We have three phases incremented on every frame (VI used as an interval)
+
+// Start: set the count register to the initial value and enter the wait loop
+// Test: VI interrupt updates the count register to the mock value
+// Timeout: If the wait loop fails to exit upon setting the mock value, the VI
+
+// interrupt will move the state into this phase, which is asserted in the test.
+// So we can test them each of them in three frames, without waiting the full
+// delay. Tested wait should be an order of magnitude longer than the VI interval
+// to prevent the loop from exiting before we are able to update the count reg.
+// If wait calculation is unable to wrap properly, the loop will run for a maximum
+// of 90 secs and endup in Timeout state, failing the tes.
+// Execution times are negligible with this setup.
+// 0x2CB4178 is 1 second
```

Comment:
Can you elaborate why this complex system with VI is required for this test? What I would do is simply to test the functions with small intervals, and then verify after the wait that the current ticks fall within a range of the expected value.

Maybe there's something I'm missing.

Reply:
I wanted to make sure to test it with intervals much larger than the execution time of any code to make sure we are actually crossing the wrapping boundary i.e it should not cross it without us knowing. Might be a little overengineered you are probably right. Should be fine with 5ms around the boundary. Let me try to do it that way and see

Reply:
Also this enables us to test a very long wait, which would be impractical o/w. Lets keep it for now if thats ok?

Reply:
OK. Out of curiosity, have you verified the test actually fails without the rest of this PR?

Reply:
Yes it does fail with the old implementation. Also, inspired by your proposal, I have changed it a little bit to actually wait for 5ms, to verify it is doing some waiting indeed :)